eth/vm/forks/constantinople/state.py

```diff
@@ -7,3 +19,10 @@
 
 class ConstantinopleState(ByzantiumState):
     computation_class = ConstantinopleComputation
+
+    def apply_transaction(self, transaction: BaseOrSpoofTransaction) -> \
+            Tuple[bytes, BaseComputation]:
+        if self.state_root != BLANK_ROOT_HASH and not self.account_db.has_root(self.state_root):
+            raise StateRootNotFound(self.state_root)
+        computation = self.execute_transaction(transaction)
+        return EMPTY_SHA3, computation
```

Comment:
This approach of sometimes returning an empty root hash seems surprising from a callers perspective, and potentially error prone. I'll look at a bigger refactor to get it done with a new API.

Reply:
> This approach of sometimes returning an empty root hash seems surprising from a callers perspective, and potentially error prone. I'll look at a bigger refactor to get it done with a new API.

I added a commit showing the direction I was imagining. If you have more you want to add, let me know.

Otherwise, I'll ask for a last review and we should be able to get it merged as is.